ui-v2/app/components/phrase-editor.js

```diff
@@ -1,44 +1,67 @@
 import Component from '@ember/component';
 import { get, set } from '@ember/object';
+import { inject as service } from '@ember/service';
 
 export default Component.extend({
+  dom: service('dom'),
   classNames: ['phrase-editor'],
   item: '',
-  remove: function(index, e) {
-    this.items.removeAt(index, 1);
-    this.onchange(e);
+  didInsertElement: function() {
+    this._super(...arguments);
+    // TODO: ref
+    this.input = get(this, 'dom').element('input', this.element);
   },
-  add: function(e) {
-    const value = get(this, 'item').trim();
-    if (value !== '') {
-      set(this, 'item', '');
-      const currentItems = get(this, 'items') || [];
-      const items = new Set(currentItems).add(value);
-      if (items.size > currentItems.length) {
-        set(this, 'items', [...items]);
-        this.onchange(e);
+  onchange: function(e) {},
+  search: function(e) {
+    // TODO: Temporarily continue supporting `searchable`
+    let searchable = get(this, 'searchable');
+    if (searchable) {
+      if (!Array.isArray(searchable)) {
+        searchable = [searchable];
       }
+      searchable.forEach(item => {
+        item.search(get(this, 'value'));
+      });
     }
+    this.onchange(e);
   },
-  onkeydown: function(e) {
-    switch (e.keyCode) {
-      case 8:
-        if (e.target.value == '' && this.items.length > 0) {
-          this.remove(this.items.length - 1);
+  oninput: function(e) {},
+  onkeydown: function(e) {},
+  actions: {
+    keydown: function(e) {
+      switch (e.keyCode) {
+        case 8: // backspace
+          if (e.target.value == '' && get(this, 'value').length > 0) {
+            this.actions.remove.bind(this)(get(this, 'value').length - 1);
+          }
+          break;
+        case 27: // escape
+          set(this, 'value', []);
+          this.search({ target: this });
+          break;
+      }
+      this.onkeydown({ target: this });
+    },
+    input: function(e) {
+      set(this, 'item', e.target.value);
+      this.oninput({ target: this });
+    },
+    remove: function(index, e) {
+      get(this, 'value').removeAt(index, 1);
+      this.search({ target: this });
+      this.input.focus();
+    },
+    add: function(e) {
+      const item = get(this, 'item').trim();
+      if (item !== '') {
+        set(this, 'item', '');
+        const currentItems = get(this, 'value') || [];
+        const items = new Set(currentItems).add(item);
```

Comment:
(thinking out loud)

Creating and throwing away a Set every time you add an item seems unfortunate, but considering it's public API, doing something where the Set is an internal implementation with an Array as the public API seems tricky when combined with the typical data-down approach.

Reply:
I think this was just for unique diff checking. Me personally, I'm not too worried about creating Sets, it's just a native object - I'm not sure if this is majorly any different from creating and throwing away Arrays?

The alternative would be loops and conditionals, Set seems to be the more appropriate tool

Reply:
Yeah, maybe I'm worried over nothing. 

To clarify, I love the use of Set here for uniquing. It's just the throwing it away on every add that raises a (probably irrelevant) flag in my head.